File: packages/sdk-swift/Sources/Rovenue/Internal/Dispatcher.swift
//  Dispatcher.swift — bridges blocking Rust-core calls into Swift async/await.
//
//  Every public async method on `Rovenue` that hits the Rust core flows through
//  this dispatcher. The Rust core's methods are synchronous (per the UniFFI
//  spec we picked in M1), so calling them on @MainActor would block the UI.
//  The dispatcher serializes them on a single user-initiated queue and
//  resumes the calling task via withCheckedThrowingContinuation.

import Foundation

internal final class Dispatcher: @unchecked Sendable {
    private let queue: DispatchQueue

    init(label: String = "dev.rovenue.sdk") {
        self.queue = DispatchQueue(label: label, qos: .userInitiated)
    }

    /// Run a throwing block on the dispatcher's serial queue and await its result.
    func run<T>(_ block: @escaping @Sendable () throws -> T) async throws -> T {
        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<T, Swift.Error>) in
            queue.async {
                do {
                    continuation.resume(returning: try block())
                } catch {
                    continuation.resume(throwing: error)
                }
            }
        }
    }

    /// Run a non-throwing block on the dispatcher's serial queue and await its result.
    func runNonThrowing<T>(_ block: @escaping @Sendable () -> T) async -> T {
        await withCheckedContinuation { (continuation: CheckedContinuation<T, Never>) in
            queue.async {
                continuation.resume(returning: block())
            }
        }
    }
}

File: packages/sdk-kotlin/src/main/kotlin/dev/rovenue/sdk/internal/Dispatcher.kt
// Dispatcher.kt — bridges blocking Rust-core calls into Kotlin suspend fun.
//
// The Rust core (per the UniFFI bindings we generate in M1) is fully
// synchronous. Calling it from the caller's coroutine context would block
// the dispatcher thread. We off-load to Dispatchers.IO, which is the
// canonical "blocking work on a thread pool" dispatcher in coroutines.
//
// Unlike the Swift façade we don't need a serial queue — the Rust core's
// own Arc<Mutex<...>> already serializes concurrent calls per-instance.
// Dispatchers.IO is a bounded thread pool (64 threads by default) optimal
// for blocking I/O.

package dev.rovenue.sdk.internal

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

internal class Dispatcher {
    suspend fun <T> run(block: () -> T): T = withContext(Dispatchers.IO) {
        block()
    }
}
